formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2017-2025 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+@file:Suppress("unused")
+@file:OptIn(ExperimentalUnsignedTypes::class)
+
+package kotlinx.serialization.cbor
+
+import kotlinx.serialization.*
+import kotlinx.serialization.cbor.internal.*
+
+/**
+ * Class representing single CBOR element.
+ * Can be [CborPrimitive], [CborMap] or [CborList].
+ *
+ * [CborElement.toString] properly prints CBOR tree as a human-readable representation.
+ * Whole hierarchy is serializable, but only when used with [Cbor] as [CborElement] is purely CBOR-specific structure
+ * which has a meaningful schemaless semantics only for CBOR.
+ *
+ * The whole hierarchy is [serializable][Serializable] only by [Cbor] format.
+ */
+@Serializable(with = CborElementSerializer::class)
+public sealed class CborElement(
+    /**
+     * CBOR tags associated with this element.
+     * Tags are optional semantic tagging of other major types (major type 6).
+     * See [RFC 8949 3.4. Tagging of Items](https://datatracker.ietf.org/doc/html/rfc8949#name-tagging-of-items).
+     */
+    @OptIn(ExperimentalUnsignedTypes::class)
+    public val tags: ULongArray = ulongArrayOf()
+)
+
+/**
+ * Class representing CBOR primitive value.
+ * CBOR primitives include numbers, strings, booleans, byte arrays and special null value [CborNull].
+ */
+@Serializable(with = CborPrimitiveSerializer::class)
+public sealed class CborPrimitive(
+    tags: ULongArray = ulongArrayOf()
+) : CborElement(tags)
+
+/**
+ * Class representing signed CBOR integer (major type 1).
+ */
+@Serializable(with = CborIntSerializer::class)
+public class CborNegativeInt(
```

Comment:
The name NegativeInt is misleading as it appears to mean a value that can only be negative, not a signed value (that can be positive or negative). Given this name is poor, you also want to rename the counterpart from PositiveInt.

Reply:
That was intentional. IIRC only negative ints are encoded as signed int in cbor, so it really must be a negative int

Reply:
But what if they are created programmatically? If you have this value based type you need a factory that creates the correct version on demand (and probably have an "CborInt" parent type that is sign independent)

Reply:
a sealed parent makes sense (with a custom invoke function that returns the subtype based on sign), but the name for this class is the only one i could think of that is actually accurate

Reply:
it definitely makes life easier when you don't parse, but create. when parsing, though, I am slightly leaning towards being very explicit and very close to the wire format (which is what we're representing). I'll let @sandwwraith decide